python-package/xgboost/core.py

```diff
@@ -959,6 +973,22 @@ def feature_types(self, feature_types: Optional[Union[List[Any], Any]]) -> None:
                 c_bst_ulong(0)))
 
 
+class _ProxyDMatrix(DMatrix):
+    """A placeholder class when DMatrix cannot be constructed (DeviceQuantileDMatrix,
+    inplace_predict).  `owner` specifies whether should the handle be freed by this
+    object.
+
+    """
+    def __init__(self, owner):  # pylint: disable=super-init-not-called
+        self.handle = ctypes.c_void_p()
+        _check_call(_LIB.XGProxyDMatrixCreate(ctypes.byref(self.handle)))
+        self._owner = owner
+
+    def __del__(self):
```

Comment:
This is dangerous. Any other way to do this?

Reply:
Let me think about it.

Reply:
Simplified it by removing some indirect functions.